readme, docs: refine language for 5 documents (#1072)
pingcap/tispark: docs/userguide.md

# TiSpark (version >= 2.0) User Guide

> **Note:**
>
> This is a user guide for TiSpark version >= 2.0. If you are using version earlier than 2.0, refer to [Document for Spark 2.1](./userguide_spark2.1.md)

This document introduces how to set up and use TiSpark, which requires some basic knowledge of Apache Spark. Refer to [Spark website](https://spark.apache.org/docs/latest/index.html) for details.

## Overview

> TiSpark is a thin layer built for running Apache Spark on top of TiDB/TiKV to answer the complex OLAP queries. It enjoys the merits of both the Spark platform and the distributed clusters of TiKV, at the same time, seamlessly integrated to TiDB, the distributed OLTP database, to provide one-stop Hybrid Transactional/Analytical Processing (HTAP) solutions for online transactions and analyses.

Reviewer: Such a long composite sentence...

Some suggestions:
- `integrated -> integrates`? If you mean by `it enjoys the merits while integrating to tidb`.
- `to provide -> provides`? If you mean by `it enjoys the merits while providing solutions`.
- Or change `at the same time` to `via/by`? If the following is considered an attributive clause.
- Or we may simply split this sentence into two for better understanding...

Author: What about `While enjoying the merits of both the Spark platform and the distributed clusters of TiKV, it is seamlessly integrated into TiDB, the distributed OLTP database, and provide one-stop Hybrid Transactional/Analytical Processing (HTAP) solutions for online transactions and analyses.`?

Reviewer: Great! Just replace `it` with `TiSpark`.

Reviewer: I'm just wondering that `providing one-stop solutions` may be a result of `integrated into TiDB`, so that they should not conjunct as parallel.

Reviewer: Also, I believe `integrate with` is better than `integrate into`. @TomShawn

Author: Yes, you are right. What about `While enjoying the merits of both the Spark platform and the distributed clusters of TiKV, it is seamlessly integrated with TiDB, the distributed OLTP database, and thus blessed to provide one-stop Hybrid Transactional/Analytical Processing (HTAP) solutions for online transactions and analyses.`?

Reviewer: nice

+ It utilizes several strategies to push down computing to reduce the size of dataset handling by Spark SQL, which accelerates query execution. It also uses the TiDB built-in statistical information for the query plan optimization.
+ From the perspective of data integration, TiSpark + TiDB provides a solution that performs both transaction and analysis directly on the same platform without building and maintaining any ETLs. It simplifies the system architecture and reduces the cost of maintenance.
+ In addition, you can deploy and utilize the tools from the Spark ecosystem for further data processing and manipulation on TiDB. For example, using TiSpark for data analysis and ETL, retrieving data from TiKV as a data source for machine learning, generating reports from the scheduling system and so on.

TiSpark relies on the availability of TiKV clusters and PDs. You also need to set up and use the Spark clustering platform.

## Prerequisites for setting up TiSpark

> + The current TiSpark version supports Spark 2.3+/2.4+, but does not support any Spark versions earlier than 2.3.

Author: What is the current TiSpark version?

Reviewer: current version of TiSpark.

> **Note:**
>
> Confirm the Spark version that your TiSpark version supports.

Suppose you already have a Spark binary, and the current PATH is `SPARKPATH`, copy the TiSpark jar package to the `$SPARKPATH/jars` directory.

#### Step 2: Start a Master node

Execute the following command on the selected Spark-Master node:

```
cd $SPARKPATH

./sbin/start-master.sh
```

After the command is executed, a log file is printed on the screen. Check the log file to confirm whether the Spark-Master is started successfully.

Open the [http://spark-master-hostname:8080](http://spark-master-hostname:8080) to view the cluster information (if you do not change the default port number of Spark-Master).

When you start Spark-Slave, you can also use this panel to confirm whether the Slave is joined to the cluster.

#### Step 3: Start a Slave node

Similarly, start a Spark-Slave node by executing the following command:

```
./sbin/start-slave.sh spark://spark-master-hostname:7077
```

> After the command returns, also see whether the Slave node is joined to the Spark cluster correctly from the panel.

Reviewer: `see -> check`?

Author: `check` is better.

|History server web UI   |  `18080`  | spark.history.ui.port  |Optional; it is only applied if you use the history server.   |
|Master port   |  `7077`  |   SPARK_MASTER_PORT  |   |
|Master REST port   |  `6066`  | spark.master.rest.port  | Not needed if you disable the `REST` service.   |
|Worker port |  (random)   |  SPARK_WORKER_PORT |   |
|Block manager port  |(random)   | spark.blockManager.port  |   |
|Shuffle server  |  `7337`   | spark.shuffle.service.port  |  Optional; it is only applied if you use the external shuffle service.  |
|  Application web UI  |  `4040`  |  spark.ui.port | If `4040` has been occupied, then `4041` is used. |

## 13. FAQ

Q: What are the pros and cons of independent deployment as opposed to a shared resource with an existing Spark / Hadoop cluster?

A: You can use the existing Spark cluster without a separate deployment, but if the existing cluster is busy, TiSpark will not be able to achieve the desired speed.

Q: Can I mix Spark with TiKV?

A: If TiDB and TiKV are overloaded and run critical online tasks, consider deploying TiSpark separately.

> You also need to consider using different NICs to ensure that OLTP's network resources are not compromised and do not affect online business.

Reviewer: ```suggestion
You also need to consider using different NICs to ensure that OLTP's network resources are not compromised so that online business is not affected.
```
